File: PolyBridge.Sample/Assets/Scripts/TimerService.cs
using System.Threading.Tasks;
using PolyBridge.Core.Attributes;
using PolyBridge.Sandbox;

[NativeService("com.polybridge.sample.timer.TimerService",
    BridgeType = typeof(TimerServiceBridge))]
[Sandbox("Timer Service")]
public partial class TimerService
{
    // 동기
    [NativeMethod("isRunning")]
    [SandboxMethod("Is Running")]
    public partial bool IsRunning();

    [NativeMethod("getTickCount")]
    [SandboxMethod("Get Tick Count")]
    public partial int GetTickCount();

    [NativeMethod("getInterval")]
    [SandboxMethod("Get Interval")]
    public partial long GetInterval();

    // 비동기
    [NativeMethod("start")]
    [SandboxMethod("Start Timer")]
    public partial Task<string> StartAsync(long intervalMs);

    [NativeMethod("stop")]
    [SandboxMethod("Stop Timer")]
    public partial Task<string> StopAsync();

    [NativeMethod("reset")]
    [SandboxMethod("Reset Timer")]
    public partial Task<string> ResetAsync();

    // Editor Mock
    [MockReturn(nameof(IsRunning))]
    internal bool MockReturnIsRunning() => false;

    [MockReturn(nameof(GetTickCount))]
    internal int MockReturnGetTickCount() => 0;

    [MockReturn(nameof(GetInterval))]
    internal long MockReturnGetInterval() => 1000L;

    [MockReturn(nameof(StartAsync))]
    internal Task<string> MockReturnStartAsync() => Task.FromResult("started");

    [MockReturn(nameof(StopAsync))]
    internal Task<string> MockReturnStopAsync() => Task.FromResult("stopped:0");

    [MockReturn(nameof(ResetAsync))]
    internal Task<string> MockReturnResetAsync() => Task.FromResult("reset:0");
}

File: PolyBridge.Sample.Android/polybridge-sample/src/main/java/com/polybridge/sample/timer/TimerService.kt
package com.polybridge.sample.timer

import java.util.Timer
import java.util.TimerTask

class TimerService {

    private var timer: Timer? = null
    private var tickCount = 0
    private var intervalMs: Long = 1000
    private var eventListener: ITimerBridge? = null

    fun setEventListener(listener: ITimerBridge) {
        eventListener = listener
    }

    fun removeEventListener(listener: ITimerBridge) {
        if (eventListener === listener) eventListener = null
    }

    // 동기
    fun isRunning(): Boolean = timer != null

    fun getTickCount(): Int = tickCount

    fun getInterval(): Long = intervalMs

    // 비동기 — 타이머 시작
    fun start(intervalMs: Long, callback: ITimerBridge) {
        if (timer != null) {
            callback.onError("Timer already running")
            return
        }

        this.intervalMs = if (intervalMs > 0) intervalMs else 1000
        tickCount = 0

        timer = Timer().apply {
            scheduleAtFixedRate(object : TimerTask() {
                override fun run() {
                    tickCount++
                    eventListener?.onTick(tickCount.toString())
                }
            }, this@TimerService.intervalMs, this@TimerService.intervalMs)
        }

        eventListener?.onTimerStateChanged("started")
        callback.onSuccess("started")
    }

    // 비동기 — 타이머 정지
    fun stop(callback: ITimerBridge) {
        if (timer == null) {
            callback.onError("Timer not running")
            return
        }

        timer?.cancel()
        timer = null
        eventListener?.onTimerStateChanged("stopped")
        callback.onSuccess("stopped:$tickCount")
    }

    // 비동기 — 리셋
    fun reset(callback: ITimerBridge) {
        timer?.cancel()
        timer = null
        val prevCount = tickCount
        tickCount = 0
        eventListener?.onTimerStateChanged("reset")
        callback.onSuccess("reset:$prevCount")
    }
}
